### PR TITLE
Reworks boiler into having to be rooted to fire

### PR DESCRIPTION
**Pull request by ivanmixo**
## About The Pull Request

Adds a new ability to boiler which makes it unable to be moved, rooting. Bombard now requires you to be rooted to use. This means that you can no longer be shuffled by your teammates, but it also means you are more vulnerable to being rushed down and bombarding a spot requires much more commitment. 

Takes 3 seconds to root and 3 seconds to unroot. You can interrupt the rooting process by moving and such.

Also cleaned up some of the absolutely schizo bombard code.

https://github.com/tgstation/TerraGov-Marine-Corps/assets/22431091/4970bf73-65f2-436f-ae4e-0aff790f3f5e
## Why It's Good For The Game

An attempt to address how non committal globbing currently is. Should help make the boiler more vulnerable while trying to bombard, be that to snipers countersniping or someone trying to rush it down.

On the flipside, it helps the boiler as well by not letting other xenos shuffle/shove the boiler while it is trying to fire.
## Changelog
:cl:
balance: Added a new ability to boiler lets the boiler "root" itself in place, unable to be moved. Being rooted is now required to bombard.
/:cl:

### CODE REVIEW

**Comment by Losenis** [comment]
>You can now bodyblock the entire hive in a maze and not even notice due to it being a smallass bunch of sprites on the edge of the screen
>You can now get completely immobilized by funny nades moving you while you're trying to deroot

I don't think you've really thought this through

**Comment by ivanmixo** [comment]
You CANNOT be moved or staggered while rooted. By anything. It's like fortify in terms of that. 

The former is a skill issue, a defender can do the same.

**Comment by Lumipharon** [comment]
This sounds like guaranteed death

**Comment by ivanmixo** [comment]
> Why not merge it with Bombard so it's only 1 ability you need to press?

It is incredibly janky. And hard to code/do right.

**Comment by Al-1ce** [comment]
I'd be fine with this if rooting was disabled after you've fired a glob. Defilers staying and shooting from the same place is generally a skill issue anyways.

**Comment by ClosetedSkeletonGH** [comment]
> This sounds like guaranteed death

boiler right now has 0 investment, if you got shot, hear arty, or have literally any risk posed to you you can just move instantly. You never need to go below 50% health as boiler. Thats the entire fault with the caste, any good boiler will never die, and its extremely easy to be a good boiler.

If adding any sort of investment means guaranteed death, the caste shouldn't exist. Every other caste has to take a risk to get damage/utility out, while boiler takes basically 0 risk.

Boiler as is right now, would not be added to the game if PR'd.

**Comment by Losenis** [comment]
>If adding any sort of investment means guaranteed death, the caste shouldn't exist.

Getting effectively stunned for 3s is pretty far in the "any sort of investment" spectrum. Even defender can get in and out of fortify immediately.

**Comment by Losenis** [comment]
I also don't think this is going to do a lot more than annoy if not straight up kill xenos given that

A) Boiler with 2 or less globs don't glow in the dark, so unless marines dunked enough points for autosniper or AMR they're still not going to be able to do much about it
B) If there's multiple boilers, even AMR or autosniper like above won't be able to do much due to chain gassing
C) If a boiler gets killed while uprooted, he's probably in a situation where he'd get killed anyway like rushers and the sort

The only niche case I can imagine making this work as intended is shooting a boiler that's already glowing and sieging fob, but that is purely skill issue on the boiler either intentionally glowing or having shot enough that people know fully well there's a boiler there instead of rotating angles.

I also don't imagine it'd make it _fun_ to play boiler as soon as a long range weapon comes into play if all it takes is a marine waiting till you uproot and there's literally nothing you can do about it once he starts shooting. AMRs are easy to spot, but even 127s can reach you and do enough damage in the 3s to unroot + ping + reaction time.

This is also super fucking janky because it's a separate ability so now you have to zoom, uproot, and bombard every time you want to shoot after moving. It's a fucking pain in the ass that I don't think anything xeno or marine has for basic actions. Even with AT post-nerf you can just hit the aim mode and zoom hotkeys at the same time and be done with it after reloading.

**Comment by Naaanii** [comment]
How is this needed in any way outside of being shuffled which is probably because the Boiler picked a bad spot to remain rooted in rather than the other way around?

Defenders can instantly unfortify while a Boiler is most likely going to wait for the whole duration till it fires. This seems like a spoonfed mechanic that's just going to backfire onto the hive for 'lack of teamwork '.

**Comment by ClosetedSkeletonGH** [comment]
> I also don't think this is going to do a lot more than annoy if not straight up kill xenos given that
> 
> A) Boiler with 2 or less globs don't glow in the dark, so unless marines dunked enough points for autosniper or AMR they're still not going to be able to do much about it B) If there's multiple boilers, even AMR or autosniper like above won't be able to do much due to chain gassing C) If a boiler gets killed while uprooted, he's probably in a situation where he'd get killed anyway like rushers and the sort
> 
> The only niche case I can imagine making this work as intended is shooting a boiler that's already glowing and sieging fob, but that is purely skill issue on the boiler either intentionally glowing or having shot enough that people know fully well there's a boiler there instead of rotating angles.
> 
> I also don't imagine it'd make it _fun_ to play boiler as soon as a long range weapon comes into play if all it takes is a marine waiting till you uproot and there's literally nothing you can do about it once he starts shooting. AMRs are easy to spot, but even 127s can reach you and do enough damage in the 3s to unroot + ping + reaction time.
> 

no it can’t lmao, SR-127, when not in aimmode and assuming that it has 0 drop off and instant bullet travel time takes 7 seconds to put a Boiler into crit assuming the boiler has 0 pheros and is not on weeds. 

This is an almost non-existent scenario because it will always be worse for the SR-127

**Comment by Losenis** [comment]
So you're telling me this pr doesn't actually do shit other than add jank for the sake of jank

Wild

**Comment by ClosetedSkeletonGH** [comment]
> So you're telling me this pr doesn't actually do shit other than add jank for the sake of jank
> 
> Wild

It makes it where the boiler has to actually put in an investment. Sure an SR-127 won’t be able to kill it at full health, but it’ll surely make it move if the boiler is playing safe (as all good boilers can do right now).

It just adds a risk to boiling

**Comment by Coastfront** [comment]
And yet we refuse to pluck an ever festering thorn.

Boiler stings into the heels of this game as fervently as it does, found just only in the thorned vines we ever choose to wade through.
The only reason boiler needs to exist is because marines have artillery mechanics.
The reason marines have artillery mechanics can seldom be explained.

In every effort to make this razor thin, tip deep thorn sting ever less it only burrows ever deeper, breaking away into countless pieces, the stars spotting its land with pain.

If boiler, is neither fun nor interactive to play against, and any effort to "balance" out their counterplay is met entirely with negative feedback because it makes the game worse, why is boiler still in the fucking game.
Nobody ever has any positive feedback about boiler nor artillery mechanics in general, It demands ever less effort than reaming a blade of grass and avails a cut away forest in return.

Boiler takes all the skill of "pressing 3 or so buttons and moving every now and again" while artillery then takes all the effort of "literally typing in a fucking number after reading the webmap" and nothing good is ever said about either.

**Comment by Naaanii** [comment]
> > So you're telling me this pr doesn't actually do shit other than add jank for the sake of jank
> > Wild
> 
> It makes it where the boiler has to actually put in an investment. Sure an SR-127 won’t be able to kill it at full health, but it’ll surely make it move if the boiler is playing safe (as all good boilers can do right now).
> 
> It just adds a risk to boiling

The risk being, pray there isn't competent marines, lol.

**Comment by ClosetedSkeletonGH** [comment]
Artillery has cool sounds, beyond that uhhhhhhhh

**Comment by Naaanii** [comment]
> And yet we refuse to pluck an ever festering thorn.
> 
> Boiler stings into the heels of this game as fervently as it does, found just only in the thorned vines we ever choose to wade through. The only reason boiler needs to exist is because marines have artillery mechanics. The reason marines have artillery mechanics can seldom be explained.
> 
> In every effort to make this razor thin, tip deep thorn sting ever less it only burrows ever deeper, breaking away into countless pieces, the stars spotting its land with pain.
> 
> If boiler, is neither fun nor interactive to play against, and any effort to "balance" out their counterplay is met entirely with negative feedback because it makes the game worse, why is boiler still in the fucking game. Nobody ever has any positive feedback about boiler nor artillery mechanics in general, It demands ever less effort than reaming a blade of grass and avails a cut away forest in return.
> 
> Boiler takes all the skill of "pressing 3 or so buttons and moving every now and again" while artillery then takes all the effort of "literally typing in a fucking number after reading the webmap" and nothing good is ever said about either.

Artillery is literally point and click, with some downtime (less downtime than boiler), just like boiler.

**Comment by Losenis** [comment]
Except you can get bonked by your own arty as marine if a baldlet or ofnir is handling it

**Review by Lumipharon — CHANGES_REQUESTED** [review]
This seems overly complicated and just adding extra buttons because ???

I'd rather just channeling a glob to fire root you, which should be like 2 lines to turn on and off.

**Comment by Naaanii** [comment]
> am i high i thought you already had to root yourself to fire?

You did but could also cancel it at any time.

File: code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm [hunk 1]
```diff
@@ -253,63 +219,109 @@ GLOBAL_LIST_INIT(boiler_glob_image_list, list(
...
+		if(boiler_owner.corrosive_ammo <= 0)
+			to_chat(boiler_owner, span_warning("We have no corrosive globules available."))
+			return FALSE
+	else
+		if(boiler_owner.neuro_ammo <= 0)
+			to_chat(boiler_owner, span_warning("We have no neurotoxin globules available."))
+			return FALSE
+
+	if(!HAS_TRAIT_FROM(boiler_owner, TRAIT_IMMOBILE, BOILER_ROOTED_TRAIT))
+		to_chat(boiler_owner, span_warning("We need to be rooted to the ground to fire!"))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm:239]: Ballon maybe?

File: code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm [hunk 2]
```diff
@@ -253,63 +219,109 @@ GLOBAL_LIST_INIT(boiler_glob_image_list, list(
...
 	update_button_icon()
 	add_cooldown()
-	X.reset_bombard_pointer()
+
+
+/datum/action/xeno_action/activable/bombard/alternate_action_activate()
+	INVOKE_ASYNC(src, PROC_REF(root))
+	return COMSIG_KB_ACTIVATED
+
+/datum/action/xeno_action/activable/bombard/proc/root()
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm:293]: Autodoc

File: code/datums/keybinding/xeno.dm [hunk 3]
```diff
@@ -224,6 +224,13 @@
 	keybind_signal = COMSIG_XENOABILITY_CREATE_BOMB
 	hotkey_keys = list("F")
 
+/datum/keybinding/xeno/root
+	name = "root"
+	full_name = "Boiler: Root in place"
+	description = ""
```
Comment (TiviPlus) [inline code/datums/keybinding/xeno.dm:230]: missing desc

File: code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm [hunk 4]
```diff
@@ -253,63 +219,109 @@ GLOBAL_LIST_INIT(boiler_glob_image_list, list(
...
+
+	owner.balloon_alert_to_viewers("Rooting into place...")
+	if(!do_after(owner, 3 SECONDS, FALSE, null, BUSY_ICON_HOSTILE))
+		owner.balloon_alert(owner, "Interrupted!")
+		return
+
+	owner.balloon_alert_to_viewers("Rooted into place!")
+	set_rooted(TRUE)
+
+/datum/action/xeno_action/activable/bombard/proc/set_rooted(on)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm:311]: ditto

**Review by TiviPlus — APPROVED** [review]

File: code/datums/keybinding/xeno.dm (hunk 3 above)
Comment (ivanmixo) [reply code/datums/keybinding/xeno.dm:230]: The majority of the binds are missing descs, I'll do the ones for boiler abilities but that's a todo right there

**Comment by Lumipharon** [comment]
I have strong concerns about this ability as is, for reasons already discussed. If it doesn't break root when your health gets too low, or atleast reduce the time or something, this is a real deathtrap.

**Comment by ivanmixo** [comment]
Doing either of those would make snipers a non threat again

**Comment by UntoldTactics** [comment]
> Doing either of those would make snipers a non threat again

Snipers have always been a threat. The majority of the boilers dominance is a marine skill issue in handling them.

**Comment by blackdav123** [comment]
to quote the smartgunner who had just soloed a boiler next to me, "it just stood there and died!!"

boiler shouldnt be forced to be an ambush caste.

if you are going to keep this then boiler could live with a massive speed buff to let it reposition quickly and always be shooting from a new spot.

**Comment by pbfied** [comment]
"An attempt to address how non committal globbing currently is. Should help make the boiler more vulnerable while trying to bombard, be that to snipers countersniping or someone trying to rush it down."

Counter sniping is also a non-committal mechanic. No one, including you, has an issue with non-committal mechanics in general. 

"On the flipside, it helps the boiler as well by not letting other xenos shuffle/shove the boiler while it is trying to fire."

You forgot to mention that this PR hurts every other Xeno who needs to traverse the boiler's space.

**Comment by BlueHNT** [comment]
It singlehandedly increased coper winrate from 50% to 60%... What else do they have left to try and change tides

**Comment by ivanmixo** [comment]
Been tmd for like 10 rounds this week I don't think it had that much of an impact

**Comment by Losenis** [comment]
You can't even drag crit boilers due to the root apparently lmao, it's an actual death sentence

**Comment by Lumipharon** [comment]
> Been tmd for like 10 rounds this week I don't think it had that much of an impact

??

**Comment by Aporhtonoma** [comment]
i have seen boilers dying because they got grenade spammed and there was actually nothing they could do to leave(other xenos could not even drag them cause funny root lol).This my dudes is not a rework is fucking nerf a very big one as well.Marines have amrs,autosnipers,obs,howis,mortars,mlrs and now tanglefoot to counter boilers and yet you all still wanna nerf them .This pr is a joke from a balance perspective if it was like warlock where it broke on stagger then maybe MAYBE it would be tolerable but right now it just makes boilers jokes and fob sieges even more of a pain than it already was.

**Comment by ivanmixo** [comment]
> > Been tmd for like 10 rounds this week I don't think it had that much of an impact
> 
> ??
> 
> edit: On a note of usability and interface QOL, its way more of a pain to actually use boiler abilities now. Previously you could just press hot key for zoom, hot key for shoot, then click. No button clicking required. Now you have to actually right click the bombard button to root and unroot, and the button moves around when you zoom, which makes it agony.

there is a hotkey use it for the love of god, do you right click the weed button when you wanna change weed type?

**Comment by ivanmixo** [comment]
> This my dudes is not a rework is fucking nerf a very big one as well

A rework can be a nerf yes

**Comment by Losenis** [comment]
>there is a hotkey use it for the love of god, do you right click the weed button when you wanna change weed type?

It's almost as if adding hotkeys on top of hotkeys that have to be used every god damn time actually affects the enjoyability of said caste regardless of power. This is on top of you not wanting to merge it with the bombard ability because it's JaNkY so you'd rather have the jank shoved to everyone else rather than pause this pr and actually take the time to learn how to do it properly.

image

**Comment by ivanmixo** [comment]
No idea what you're on about. It literally is one ability now and that comment is outdated.

Issue is prefs are shit and don't automatically update your hotkeys to a default one when a hotkey gets added, and there is no indication that an ability has an alternate ability which is very bleh.

**Comment by Mannybrado** [comment]
another ridiculous nerf presented as a "rework"

**Comment by MiniMeatwad** [comment]
Add an auto un-root when you press a movement key, it will save me who lags for 20 seconds when I use long-range vision. Because I can't use an ability while lagging, but I can hold down a movement key and hope it ejects me in a general direction

**Pull request merged**